src/proof/lfsc/lfsc_printer.cpp

```diff
@@ -0,0 +1,515 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The printer for LFSC proofs
+ */
+
+#include "proof/lfsc/lfsc_printer.h"
+
+#include <sstream>
+
+#include "expr/dtype.h"
+#include "expr/dtype_cons.h"
+#include "expr/dtype_selector.h"
+#include "expr/node_algorithm.h"
+#include "expr/skolem_manager.h"
+#include "proof/lfsc/lfsc_list_sc_node_converter.h"
+#include "proof/lfsc/lfsc_print_channel.h"
+
+using namespace cvc5::kind;
+
+namespace cvc5 {
+namespace proof {
+
+LfscPrinter::LfscPrinter(LfscNodeConverter& ltp)
+    : d_tproc(ltp), d_assumpCounter(0)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  d_boolType = nm->booleanType();
+  // used for the `flag` type in LFSC
+  d_tt = d_tproc.mkInternalSymbol("tt", d_boolType);
+  d_ff = d_tproc.mkInternalSymbol("ff", d_boolType);
+}
+
+void LfscPrinter::print(std::ostream& out,
+                        const std::vector<Node>& assertions,
+                        const ProofNode* pn)
+{
+  // to be added
+}
+
+void LfscPrinter::printProofLetify(
+    LfscPrintChannel* out,
+    const ProofNode* pn,
+    const LetBinding& lbind,
+    const std::vector<const ProofNode*>& pletList,
+    std::map<const ProofNode*, size_t>& pletMap,
+    std::map<Node, size_t>& passumeMap)
+{
+  // closing parentheses
+  size_t cparen = 0;
+
+  // define the let proofs
+  if (!pletList.empty())
+  {
+    std::map<const ProofNode*, size_t>::iterator itp;
+    for (const ProofNode* p : pletList)
+    {
+      itp = pletMap.find(p);
+      Assert(itp != pletMap.end());
+      size_t pid = itp->second;
+      // print (plet _ _
+      out->printOpenLfscRule(LfscRule::PLET);
+      cparen++;
+      out->printHole();
+      out->printHole();
+      out->printEndLine();
+      // print the letified proof
+      pletMap.erase(p);
+      printProofInternal(out, p, lbind, pletMap, passumeMap);
+      pletMap[p] = pid;
+      // print the lambda (\ __pX
+      out->printOpenLfscRule(LfscRule::LAMBDA);
+      cparen++;
+      out->printProofId(pid);
+      out->printEndLine();
+    }
+    out->printEndLine();
+  }
+
+  // [2] print the proof body
+  printProofInternal(out, pn, lbind, pletMap, passumeMap);
+
+  // print the closing parenthesis
+  out->printCloseRule(cparen);
+}
+
+void LfscPrinter::printProofInternal(
+    LfscPrintChannel* out,
+    const ProofNode* pn,
+    const LetBinding& lbind,
+    const std::map<const ProofNode*, size_t>& pletMap,
+    std::map<Node, size_t>& passumeMap)
+{
+  // the stack
```

Comment:
Just making sure, you introduced a utility for such traversal / visits not so long ago. Can't it be used here?

Reply:
This is a new kind of traversal, one where the next things to print can either be a proof or a term (or a sort).